src/diepvries/template_sql/effectivity_satellite_dml.sql

```diff
@@ -1,78 +1,84 @@
-          INNER JOIN {target_schema}.{link_table} AS l
-                     ON ({link_driving_key_condition})
-          INNER JOIN {target_schema}.{target_table} AS satellite
-                     ON (l.{hashkey_field} = satellite.{hashkey_field}
+        WITH
+          effectivity_satellite AS (
+          SELECT
+            {link_driving_keys},
+            satellite.*
+          FROM {target_schema}.{link_table} AS l
+            INNER JOIN {target_schema}.{target_table} AS satellite
+                       ON (l.{hashkey_field} = satellite.{hashkey_field}
                          AND satellite.{record_end_timestamp_name} = {end_of_time})
-      ),
-      staging_satellite_affected_records AS (
-        /* Records that will be inserted (don't exist in target table or exist
-          in the target table but the hashdiff changed). As the r_timestamp is fetched
-          from the staging table, these records will always be included in the
-          WHEN NOT MATCHED condition of the MERGE command. */
-        SELECT
-          {staging_driving_keys},
-          staging.{hashkey_field},
-          staging.{staging_hashdiff_field},
-          staging.{record_start_timestamp},
-          staging.{record_source}
-          {staging_descriptive_fields}
-        FROM filtered_staging AS staging
-          LEFT JOIN effectivity_satellite AS satellite
-                    ON ({satellite_driving_key_condition})
-        WHERE satellite.{hashkey_field} IS NULL
-           OR satellite.{hashdiff_field} <> staging.{staging_hashdiff_field}
-        UNION ALL
-        /* Records from the target table that will have its r_timestamp_end updated
-          (hashkey already exists in target table, but hashdiff changed). As the
-          r_timestamp is fetched from the target table, these records will always be
-          included in the WHEN MATCHED condition of the MERGE command. */
+                                   ),
+          filtered_effectivity_satellite AS (
+          SELECT
+            satellite.*
+          FROM {staging_schema}.{staging_table} AS staging
+            INNER JOIN effectivity_satellite AS satellite
+                       ON ({satellite_driving_key_condition})
+                                            ),
+          filtered_staging AS (
+          SELECT DISTINCT
+            {staging_driving_keys},
+            staging.{hashkey_field},
+            staging.{staging_hashdiff_field},
+            staging.{record_start_timestamp},
+            staging.{record_source}
+            {staging_descriptive_fields}
+          FROM {staging_schema}.{staging_table} AS staging
+          WHERE NOT EXISTS (
+                           SELECT
+                             1
+                           FROM filtered_effectivity_satellite AS satellite
+                           WHERE {satellite_driving_key_condition}
+                            AND satellite.r_timestamp >= staging.r_timestamp
+                           )
```

Comment:
Instead of looking at all the records in the satellite and retrieving the max timestamp, we look only at the keys which are included in the staging table, right?

Reply:
Yup! Same logic but at a more granular level.